File: pyzoo/zoo/pipeline/api/net/torch_criterion.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import torch
import torch.nn as nn
import sys
import os
import tempfile
import shutil
from bigdl.nn.criterion import Criterion
from .torch_net import TorchNet

if sys.version >= '3':
    long = int
    unicode = str


class LossWrapperModule(nn.Module):
    def __init__(self, lossFunc):
        super(LossWrapperModule, self).__init__()
        self.func = lossFunc

    def forward(self, x, y):
        return self.func(x, y)


class TorchCriterion(Criterion):
    """
    TorchCriterion wraps a loss function for distributed inference or training.
    Use TorchCriterion.from_pytorch to initialize.
    """

    def __init__(self, path, bigdl_type="float"):
        """
        :param path: path to the TorchScript model.
        :param bigdl_type:
        """
        super(TorchCriterion, self).__init__(None, bigdl_type, path)

    @staticmethod
    def from_pytorch(loss, input_shape=None, label_shape=None,
                     sample_input=None, sample_label=None):
        """
        Create a TorchCriterion directly from PyTorch function. We need user to provide a sample
        input and label to trace the loss function. User may just specify the input and label shape.
        For specific data type or multiple input models, users can send sample_input and
        sample_label.
        :param loss: this can be a torch loss (e.g. nn.MSELoss()) or
                     a function that take two Tensor parameters: input and label. E.g.
                     def lossFunc(input, target):
                         return nn.CrossEntropyLoss().forward(input, target.flatten().long())

        :param input_shape: list of integers.
        :param label_shape: list of integers. If not specified, it will be set equal to input_shape
        :param sample_input: a sample of input.
        :param sample_label: a sample of label.
        """
        if not input_shape and not label_shape and not sample_input and not sample_label:
            raise Exception("please specify input_shape and label_shape, or sample_input"
                            " and sample_label")

        temp = tempfile.mkdtemp()

        # use input_shape as label shape when label_shape is not specified
        if not label_shape:
            label_shape = input_shape

        sample_input = TorchNet.get_sample_input(input_shape, sample_input)
        sample_label = TorchNet.get_sample_input(label_shape, sample_label)

        traced_script_loss = torch.jit.trace(LossWrapperModule(loss),
                                             (sample_input, sample_label))
        lossPath = os.path.join(temp, "loss.pt")
        traced_script_loss.save(lossPath)

        criterion = TorchCriterion(lossPath)
        shutil.rmtree(temp)

        return criterion

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TorchCriterion.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intel.analytics.zoo.pipeline.api.net

import java.io.{File, IOException}
import java.nio.file.{Files, Paths}

import com.intel.analytics.bigdl.Module
import com.intel.analytics.bigdl.nn.abstractnn.{AbstractCriterion, Activity}
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.{T, Table}
import com.intel.analytics.zoo.pipeline.api.net.TorchNet.TorchModelHolder
import org.apache.commons.io.FileUtils

import scala.reflect.ClassTag

class TorchCriterion private(private val lossHolder: TorchModelHolder)
    extends AbstractCriterion[Activity, Activity, Float]() {

  implicit val ev = TensorNumeric.NumericFloat
  implicit val tag: ClassTag[Float] = ClassTag.Float

  /**
   * sequential id in cpp: std::vector<std::shared_ptr<torch::jit::script::Module>> handles;
   * mark the model as transient and reload TorchNet from byteArray on executors
   */
  @transient
  lazy val nativeRef: Long = {
    println("TorchCriterion loading in " + this)
    val ref = TorchCriterion.loadPytorchCriterion(lossHolder.torchBytes)
    ref
  }

  override def updateOutput(input: Activity, target: Activity): Float = {
    val inputTabel = if (input.isTensor) T(input.toTensor) else input.toTable
    val targetTable = if (target.isTensor) T(target.toTensor) else target.toTable

    val (sto1, off1, shape1) = TorchCriterion.extract(inputTabel)
    val (sto2, off2, shape2) = TorchCriterion.extract(targetTable)

    val result = PytorchModelWrapper.lossForwardNative(nativeRef, sto1, off1,
      shape1, sto2, off2, shape2)
    Tensor(result.getData, result.getShape).mean()
  }

  override def updateGradInput(input: Activity, target: Activity): Activity = {

    val result = PytorchModelWrapper.lossBackwardNative(nativeRef)
    if (result.length == 1) {
      val resultTensor = Tensor(result(0).getData, result(0).getShape)
      if (gradInput == null) {
        gradInput = Tensor()
      }
      gradInput.toTensor.set(resultTensor)
    } else {
      if (gradInput == null) {
        gradInput = T()
      }
      gradInput.toTable.clear()
      result.foreach { t =>
        gradInput.toTable.insert(Tensor(t.getData, t.getShape))
      }
    }
    gradInput
  }

  override def finalize(): Unit = {
    super.finalize()
    PytorchModel.releaseLossNative(nativeRef)
  }

}

object TorchCriterion {

  /**
   * Create a TorchCriterion from a saved TorchScript Model
   * @param lossPath Path to the TorchScript Model.
   * @return
   */
  def apply(lossPath: String): TorchCriterion = {
    // TODO: add support for HDFS path
    val modelbytes = Files.readAllBytes(Paths.get(lossPath))
    new TorchCriterion(new TorchModelHolder(modelbytes, lossPath))
  }

  private[net] def loadPytorchCriterion(bytes: Array[Byte]): Long = {
    var nativeRef = -1L
    try {
      val tmpFile = File.createTempFile("TorchNet", "_pt")
      Files.write(Paths.get(tmpFile.toURI), bytes)
      nativeRef = PytorchModel.loadLossNative(tmpFile.getAbsolutePath)
      FileUtils.deleteQuietly(tmpFile)
    } catch {
      case io: IOException =>
        System.out.println("error during loading Torch model")
        throw io
    }
    nativeRef
  }


  private[net] def extract(t: Table): (Array[Array[Float]], Array[Int], Array[Array[Int]]) = {
    val tensors = t.toSeq[Tensor[Float]]

    val tuples = tensors.map { t =>
      require(t.isContiguous())
      (t.storage(), t.storageOffset() - 1, t.size())
    }
    val storages = tuples.map(_._1.array()).toArray
    val offsets = tuples.map(_._2).toArray
    val shapes = tuples.map(_._3).toArray
    (storages, offsets, shapes)
  }

}
